### DISCUSSION TITLE
Vocabulary base class / interface etc

### DISCUSSION THREAD

**Discussion by barthanssens** [eclipse-rdf4j/rdf4j#3188] [Ideas]
Maybe over-engineering / over-thinking this a bit:

- new vocabulary abstract class or interface
- methods getNamespace(), getPrefix(), getClasses(), getProperties(), getIndividuals()
- separate packages for 
   - W3C: (almost) all of current vocabularies
   - Schema.org (if needed)
   - EU Core Vocabularies (perhaps a bit niche, but I'd be interested in using these)
- "official" tool to generate Java class from OWL

**Comment by abrokenjester** [comment]
We have had several attempts to create some abstract class or interface for vocabularies in the past, including a design using enums, but they all had various downsides and were never really adopted (see a.o. https://openrdf.atlassian.net/browse/SES-1850). One problem is that static constants and inheritance don't really mix well, so you often end up with an abstract class that gives a template, but then every concrete vocabulary just reimplements/hides the whole lot again to make it work. 

That being said, happy to discuss a fresh design attempt. Perhaps more modern Java techniques open up new options.

**Comment by abrokenjester** [comment]
As for a tool to generate Java classes from OWL: I assume you are "merely" talking about creating vocabulary constants here, and not about some sort of OO mapper?

**Comment by hmottestad** [comment]
If we are introducing new packages I think we should consider also moving away from the ALL_CAPS naming scheme. So we can have RDF.type and RDF.Property. I don't think we should go as far as to lowercase the class names, e.g rdf.type.

**Comment by barthanssens** [comment]
> As for a tool to generate Java classes from OWL: I assume you are "merely" talking about creating vocabulary constants here, and not about some sort of OO mapper?

Correct, just creating a Java class (using a template engine or just string concatenation) with final static URIs for properties and classes, not an OO mapper

**Comment by barthanssens** [comment]
> If we are introducing new packages I think we should consider also moving away from the ALL_CAPS naming scheme. So we can have RDF.type and RDF.Property.

It would help for vocabularies that have similar named properties and classes (which is a questionable practice IMHO), e.g. DCAT with dcat:dataset (property) and dcat:Dataset (class).

Currently we have to use something like HAS_DATASET and DATASET to avoid name clashes...

> I don't think we should go as far as to lowercase the class names, e.g rdf.type.

Indeed, that would look a bit... weird ...

**Comment by barthanssens** [comment]
The vocabulary generator could (also) be part of the Console tool

**Comment by vorburger** [comment]
#5284 created related to this.

#2265 was original (now locked) related issue; I've only found this discussion after creating that issue - my apologies.
